### PR TITLE
loop customization docs 

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Continuation of #9557 with new intro. 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/advanced/loops.rst [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+This enables a great amount of control when needed, but the loop structure here still remains predetermined and
+fully controlled by the Trainer.
+
+Loop customization enables a new level of control where also the two remaining for loops and more can be fully changed or replaced.
+
+Here is how the above training loop can be defined using the new Loop API:
+
+.. code-block:: python
+
+    class MainLoop(Loop):
```
Comment (tchaton) [inline docs/source/advanced/loops.rst:61]:
```suggestion
    class FitLoop(Loop):
```

File: docs/source/advanced/loops.rst [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            self.iterator = enumerate(dataloader)
+            while True:
+                try:
+                    self.advance()
+                except StopIteration:
+                    break
+
+        def advance(self):
+            i, batch = next(self.iterator)
+            lightning_module.training_step(i, batch)
```
Comment (tchaton) [inline docs/source/advanced/loops.rst:87]:
```suggestion
            loss = lightning_module.training_step(batch, i)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
```

File: docs/source/extensions/loops.rst [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+.. code-block:: python
+
+    for i, batch in enumerate(dataloader):
+        loss = lightning_module.training_step(batch, i)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+What remains in the Trainer is the loop, :code:`zero_grad()`, :code:`backward()` and :code:`optimizer.step()` calls.
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:40]:
```suggestion
The other operations are automated by the Lightning Trainer: :code:`zero_grad()`, :code:`backward()` and :code:`optimizer.step()` calls.
```

File: docs/source/extensions/loops.rst [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+These are considered *boilerplate* and get automated by Lightning.
+
+This optimization scheme is very general and applies to the vast majority of deep learning research today.
+However, the loops and optimizer calls here remain predetermined in their order and are fully controlled by the Trainer.
+
+Loop customization now enables a new level of control where also the two remaining `for` loops and more can be fully changed or replaced.
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:46]:
```suggestion
With the new `Loop` class, you can now fully customize the optimization scheme for any use case.
```

File: docs/source/extensions/loops.rst [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+            while True:
+                try:
+                    self.advance()
+                except StopIteration:
+                    break
+
+Defining a loop with a class interface instead of hard-coding a raw Python for/while loop has several benefits:
+
+1. you can have full control over the data flow through loops
+2. you can add new loops and nest as many of them as they want
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:71]:
```suggestion
2. you can add new loops and nest as many of them as you want
```

File: docs/source/extensions/loops.rst [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+            self.iterator = enumerate(dataloader)
+            while True:
+                try:
+                    self.advance()
+                except StopIteration:
+                    break
+
+Defining a loop with a class interface instead of hard-coding a raw Python for/while loop has several benefits:
+
+1. you can have full control over the data flow through loops
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:70]: ### 
```suggestion
1. You can have full control over the data flow through loops
```

File: docs/source/extensions/loops.rst [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+2. you can add new loops and nest as many of them as they want
+3. if needed, the state of a loop can be saved and resumed
+4. new hooks can be injected at any point
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:73]:
```suggestion
4. New hooks can be injected at any point
```

File: docs/source/extensions/loops.rst [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+
+Loop customization is an experimental feature introduced in Lightning 1.5 that enables users to swap the built-in Lightning loops with custom ones, providing a unprecedented level of flexibility.
+Until now, Lightning did not support well some training paradigms like recommendation engine optimization or active learning.
+The loop customization feature will not only enable researchers to customize Lightning down to its very core, but also allow one to build new functionalities on top of it.
+
+.. image:: ../_static/images/extensions/loops/epoch-loop-steps.gif
+    :alt: Animation showing how to convert a standard training loop to a Lightning loop
+
+
+The training loop in Lightning
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:14]:
```suggestion
The Built-in Training Loop
```

File: docs/source/extensions/loops.rst [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+What remains in the Trainer is the loop, :code:`zero_grad()`, :code:`backward()` and :code:`optimizer.step()` calls.
+These are considered *boilerplate* and get automated by Lightning.
+
+This optimization scheme is very general and applies to the vast majority of deep learning research today.
+However, the loops and optimizer calls here remain predetermined in their order and are fully controlled by the Trainer.
+
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:45]: Add a subtitle here

Customized Lightning Loops
----------------------------

File: docs/source/extensions/loops.rst [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+     - :class:`~pytorch_lightning.loops.dataloader.evaluation_loop.EvaluationLoop`
+   * - :meth:`~pytorch_lightning.trainer.trainer.Trainer.predict`
+     - :attr:`~pytorch_lightning.trainer.trainer.Trainer.predict_loop`
+     - :class:`~pytorch_lightning.loops.dataloader.prediction_loop.PredictionLoop`
+
+
+When the user calls :code:`Trainer.<entry-point>`, it redirects to the corresponding :code:`Trainer.loop.run()` which implements the main logic of that particular Lightning loop.
+The :meth:`~pytorch_lightning.loops.base.Loop.run` method is part of the base :class:`~pytorch_lightning.loops.base.Loop` class that every loop inherits from (like every model inherits from LightningModule).
+
+Adding a custom loop for any of these entry points is straightforward.
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:108]:
```suggestion
Customizing any of these entry point loops is simple:
```

File: docs/source/extensions/loops.rst [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+
+**Step 1:** Subclass one of the above loop classes.
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:110]:
```suggestion
**Step 1:** Subclass one of the above loop classes (or inherit `Loop` to start from scratch)
```

File: docs/source/extensions/loops.rst [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+
+Given this new loop definition, here is how you connect it to the :code:`Trainer`:
+
+.. code-block:: python
+
+    model = LitModel()
+    trainer = Trainer()
+
+    yield_loop = YieldLoop()
+    trainer.fit_loop.epoch_loop.batch_loop.connect(optimizer_loop=yield_loop)
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:256]: This is different than the explanation at the top, where you just did trainer.fit_loop(loop).

Can you explain why this is required? Or have a more basic example?

File: docs/source/extensions/loops.rst [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+    trainer.fit(model)  # runs the new loop!
+
+Note that we need to connect it to the :class:`~pytorch_lightning.loops.batch.training_batch_loop.TrainingBatchLoop` as this is the next higher loop above the optimizer loop.
```
Comment (edenlightning) [inline docs/source/extensions/loops.rst:259]: This will be better as a comment in the code + explain this before the code snippet.

**Review by edenlightning — COMMENTED** [review]
Awesome work!!! This is so cool.

I suggest the following structure to the docs

Loops
====

Control every part of the training loop with the new Lightning Loop API
Why loops?
-----------
The gist: Some use cases require different training technique that isn't well supported in lightning
* Active learning
* Recommender optizations 
* etc

Lightning Trainer built-in loops
-------------------------------

What the base loop is like
Explain the entry points, etc

Override the trainer loops
-------------------------
To customize the training loop, simply
1. override loop (+ code snippet)
2. bla bla (code snippet)

Example
--------
Show basic example end to end
explain motivation for the example

code snippet

Advanced example
------------------
How to create a loop from scratch
motivation

code snippet

File: docs/source/extensions/loops.rst [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+     - :class:`~pytorch_lightning.loops.fit_loop.FitLoop`
+   * - :meth:`~pytorch_lightning.trainer.trainer.Trainer.validate`
+     - :attr:`~pytorch_lightning.trainer.trainer.Trainer.validate_loop`
+     - :class:`~pytorch_lightning.loops.dataloader.evaluation_loop.EvaluationLoop`
+   * - :meth:`~pytorch_lightning.trainer.trainer.Trainer.test`
+     - :attr:`~pytorch_lightning.trainer.trainer.Trainer.test_loop`
+     - :class:`~pytorch_lightning.loops.dataloader.evaluation_loop.EvaluationLoop`
+   * - :meth:`~pytorch_lightning.trainer.trainer.Trainer.predict`
+     - :attr:`~pytorch_lightning.trainer.trainer.Trainer.predict_loop`
+     - :class:`~pytorch_lightning.loops.dataloader.prediction_loop.PredictionLoop`
```
Comment (tchaton) [inline docs/source/extensions/loops.rst:102]: I really liked the FitLoop being decomposed as before into its part. I believe we might want to keep all this but at the bottom of the page.

More Technical Section.
...

File: docs/source/extensions/loops.rst [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+.. code-block:: python
+
+    def run(self, *args, **kwargs):
+
+        self.reset()
+        self.on_run_start(*args, **kwargs)
+
+        while not self.done:
+            try:
+                self.advance(*args, **kwargs)
```
Comment (tchaton) [inline docs/source/extensions/loops.rst:309]: Should we add advance_start, advance_end ?

File: docs/source/extensions/loops.rst [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+    # or
+    trainer.predict_loop = loop
+    trainer.predict(model)
+
+
+Practical example: the training step as a generator
+---------------------------------------------------
+
+Lightning supports multiple optimizers and offers a special :code:`training_step` flavor for it, where an extra argument with the current optimizer being used is passed in.
+Take as an example the following training step of a DCGAN from the Lightning Bolts repository:
```
Comment (tchaton) [inline docs/source/extensions/loops.rst:149]:
```suggestion
Take as an example the following training step of a DCGAN from the `Lightning Bolts <https://github.com/PyTorchLightning/lightning-bolts/>`_ repository:
```

File: docs/source/extensions/loops.rst [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+
+    # 3.3 Finetune
+    trainer.finetune(model, datamodule=datamodule, strategy="freeze")
+
+    # 4. Predict what's on a few images! ants or bees?
+    predictions = model.predict("data/hymenoptera_data/val/bees/65038344_52a45d090d.jpg")
+    print(predictions)
+
+    # 5. Save the model!
+    trainer.save_checkpoint("image_classification_model.pt")
```
Comment (tchaton) [inline docs/source/extensions/loops.rst:382]:
```suggestion
    trainer.save_checkpoint("image_classification_model.pt")
    
Here is the `example <https://github.com/PyTorchLightning/lightning-flash/blob/master/flash_examples/integrations/baal/image_classification_active_learning.py>`_.  

Here is the associated `code <https://github.com/PyTorchLightning/lightning-flash/blob/master/flash/image/classification/integrations/baal/loop.py#L31>`_.  
```

**Review by tchaton — COMMENTED** [review]
Looks great !

I would add an additional section `For advanced users`.

From the v1 PR, I would keep all the parts about all the nested loops and move it at the bottom of this one under a disclaimer this is more advanced.
